ouroboros-network/src/Ouroboros/Network/Protocol/Trans/Hello/Codec.hs

```diff
@@ -0,0 +1,180 @@
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Read as CBOR
+import           Text.Printf
+
+
+byteLimitsHello
+    :: forall ps (stIdle :: ps) bytes .
+       ProtocolSizeLimits ps bytes
+    -> ProtocolSizeLimits (Hello ps stIdle) bytes
+byteLimitsHello ProtocolSizeLimits { sizeLimitForState, dataSize }
+    = ProtocolSizeLimits {
+          sizeLimitForState = sizeLimitForStateHello,
+          dataSize
+        }
+  where 
+    sizeLimitForStateHello :: forall (pr :: PeerRole) (st :: Hello ps stIdle).
+                              PeerHasAgency pr st -> Word
+    sizeLimitForStateHello (ClientAgency TokIdle) =
+      smallByteLimit
+    sizeLimitForStateHello (ClientAgency (TokClientWrapped tok)) =
+      sizeLimitForState (ClientAgency tok)
+    sizeLimitForStateHello (ServerAgency (TokServerWrapped tok)) =
+      sizeLimitForState (ServerAgency tok)
+
+
+timeLimitsHello
+    :: forall ps (stIdle :: ps).
+       ProtocolTimeLimits ps
+    -> ProtocolTimeLimits (Hello ps stIdle)
+timeLimitsHello ProtocolTimeLimits { timeLimitForState } =
+    ProtocolTimeLimits {
+        timeLimitForState = timeLimitForStateHello
+      }
+  where
+    timeLimitForStateHello :: forall (pr :: PeerRole) (st :: Hello ps stIdle).
+                               PeerHasAgency pr st -> Maybe DiffTime
+    timeLimitForStateHello (ClientAgency TokIdle) = waitForever
+    timeLimitForStateHello (ClientAgency (TokClientWrapped tok)) =
+      timeLimitForState (ClientAgency tok)
+    timeLimitForStateHello (ServerAgency (TokServerWrapped tok)) =
+      timeLimitForState (ServerAgency tok)
+
+
+codecHello
+    :: forall ps (stIdle :: ps) m.
+       ( MonadST m
+       , ShowProxy ps
+       , ShowProxy stIdle
+       , forall (st :: ps). Show (ClientHasAgency st)
+       , forall (st :: ps). Show (ServerHasAgency st)
+       )
+    => (forall (pr :: PeerRole) (st :: ps) (st' :: ps).
+               PeerHasAgency pr st
+            -> Message ps st st' -> CBOR.Encoding)
+    -> (forall (pr :: PeerRole) (st :: ps) s.
+               PeerHasAgency pr st
+            -> CBOR.Decoder s (SomeMessage st))
+    -> Codec (Hello ps stIdle) CBOR.DeserialiseFailure m LBS.ByteString
+codecHello encode decode = mkCodecCborLazyBS encodeHello decodeHello
+  where
+    encodeHello :: forall (pr  :: PeerRole)
+                     (st  :: Hello ps stIdle)
+                     (st' :: Hello ps stIdle).
+                  PeerHasAgency pr st
+               -> Message (Hello ps stIdle) st st'
+               -> CBOR.Encoding
+    encodeHello (ClientAgency TokIdle) MsgHello =
+      encodeListLen 1 <> encodeWord 0
+    encodeHello (ClientAgency (TokClientWrapped tok)) (MsgWrapped msg) =
+         encodeListLen 2
+      <> encodeWord 1
+      <> encode (ClientAgency tok) msg
+    encodeHello (ServerAgency (TokServerWrapped tok)) (MsgWrapped msg) =
+         encodeListLen 2
+      <> encodeWord 2
+      <> encode (ServerAgency tok) msg
```

Comment:
I really don't think we want to do this.

We can make this hello extension almost completely transparent. We do not need to add extra encoding overhead to every single later message.

I see that your worry is that the initial hello message might clash with the encoding of later messages, but that cannot happen! We can take advantage of the fact that messages from different protocol states can never happen at the same time and so can never clash or be confused with each other.

So we should use exactly the underlying encoder for the wrapped states, and just pick anything for the hello message. It could be 0, or [] or null, or "hello". I'd go for null since we tend not to use that elsewhere in protocol codecs.

Reply:
You're right, it's more general than it needs to be for our purpose.

Reply:
Yet, a better solution is to have a function `Int -> Codec` in `Hello`, and define the tag in `TxSubmission2.Codec` module.  This way we can extend the `TxSubmission` with the next  tag, and once we will inline `Hello`, `TxSubmission` will look good too, without any scars of the past.